### PR TITLE
Sentry Integration

### LINKED ISSUE #207: Sentry integration
Much like wrangler1, we should have error tracking via Sentry. Assigning to @JacobMGEvans as per discussion yesterday. 

- [ ] - implement user confirmation
- [ ] - test sentry token
- [ ] - implement error capture

### PR DESCRIPTION
**Pull request by JacobMGEvans**
Sentry will provide insights to errors and debugging potentially allow us to catch bugs before they are even reported in GH issues. It is also inline with matching Wrangler 1 in functionality. 

- Top level unhandled exception logging
- Source map "inline" for typescript sourcing for Sentry
- TypeScript support with RewriteFrames config

Closes #207

### CODE REVIEW

File: packages/wrangler/src/cli.ts
Comment (threepointone) [inline packages/wrangler/src/cli.ts:9]: This will leak user data. Is there a better option? 

Also, why are we attaching it to global? It appears to be directly referenced in this file, so can't it be a regular let/const?

Comment (JacobMGEvans) [reply packages/wrangler/src/cli.ts:9]: Seemed like it was necessary for TypeScript, I can try running some test runs without it. This is what I was basing it off https://docs.sentry.io/platforms/node/typescript/#changing-events-frames

Comment (JacobMGEvans) [reply packages/wrangler/src/cli.ts:9]: Seems like I can still see the Source just fine without it. I will drop that line.
Screen Shot 2022-01-11 at 10 51 38 AM

Comment (JacobMGEvans) [reply packages/wrangler/src/cli.ts]: removed the `RewriteFrames` integration which was referencing the "global rootdir"

**Comment by threepointone** [comment]
I would love to see a demo of this before landing this. Could you do a demo for us some time?

**Comment by JacobMGEvans** [comment]
> I would love to see a demo of this before landing this. Could you do a demo for us some time?

Absolutely. I still have my personal Sentry setup.

File: packages/wrangler/src/cli.ts
Comment (Electroid) [inline packages/wrangler/src/cli.ts:11]:
```suggestion
    tags: { wrangler: pkj.version },
```

Comment (Electroid) [reply packages/wrangler/src/cli.ts:11]: Could we also add some metadata? OS, architecture, Node version, etc. (Account ID would be great to identify issues as well, but you would need to add that somewhere in `cfetch`, I think)

Comment (JacobMGEvans) [reply packages/wrangler/src/cli.ts:11]: In local changes I had already changed it to be the same `pkj.name` as the release. As for making it `wrangler` wouldn't that make it harder to distinguish from older wrangler when running in the same "Wrangler" project in Sentry. 

As for metadata and any other additional insights, we can definitely look into that 😄

**Comment by JacobMGEvans** [comment]
@threepointone I was able to get the sourcemap working. Tested localling with the sentry-cli and personal Sentry. 

The next steps will be adding uploading the sourcemaps during the build step then removing them. 
Screen Shot 2022-01-15 at 1 48 25 PM

File: packages/wrangler/package-lock.json
Comment (petebacondarwin) [inline packages/wrangler/package-lock.json]: Looks like you ran `npm install` inside the `packages/wrangler` directory.
Please can you delete this file.
The correct way to add a dependency to wrangler is to run `npm` from the top level and target its workspace:

```sh
npm install -w wrangler @sentry/node
```

File: packages/wrangler/scripts/bundle.ts
Comment (petebacondarwin) [inline packages/wrangler/scripts/bundle.ts:25]: If esbuild is set to `sourcemap: external` I am surprised that it does not automatically add this comment to the output...

Comment (petebacondarwin) [reply packages/wrangler/scripts/bundle.ts:25]: Oh I see "external" explicitly does not add the comment. I think we want "linked" here?
https://esbuild.github.io/api/#sourcemap

Comment (petebacondarwin) [reply packages/wrangler/scripts/bundle.ts:25]: Indeed it looks like the default when `sourcemap: true` is "external"
https://github.com/evanw/esbuild/blob/master/lib/shared/common.ts#L360

File: packages/wrangler/src/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/cli.ts:25]: Do we ever get an unhandled exception? Below we have a `.catch()` handler on the call to `main()`.

File: packages/wrangler/src/cli.ts
Comment (petebacondarwin) [inline packages/wrangler/src/cli.ts:14]: Should this URL be stored in the code for public consumption? This could allow people to send fake data to our logging or try to overload it...

If we got esbuild to read it from the environment at build time then we could hide it in our CI.

**Review by petebacondarwin — COMMENTED** [review]
Thanks for putting this together @JacobMGEvans - I would also like to see a demo of the result of the data in Sentry.

File: packages/wrangler/package.json
Comment (threepointone) [inline packages/wrangler/package.json:39]: These should be in devDependencies

File: packages/wrangler/scripts/bundle.ts
Comment (threepointone) [reply packages/wrangler/scripts/bundle.ts:25]: I'm certain esbuild adds the line with `sourcemap:true`, so this addition shouldn't be necessary.

File: packages/wrangler/src/cli.ts
Comment (threepointone) [reply packages/wrangler/src/cli.ts:14]: They'd still be able to see it in the bundle. This is fine really. 
What we do need before we land this PR is to figure out an opt-out system.

File: packages/wrangler/src/cli.ts
Comment (threepointone) [inline packages/wrangler/src/cli.ts:32]: I'm still concerned that the demo we saw didn't crash the process, I think we should investigate that further.

**Review by threepointone — COMMENTED** [review]
We need to figure out how we're going to do opt-out. It can't just be an environment var, we may have to add it in ~/.wrangler or something. Let's discuss this further.

File: packages/wrangler/src/cli.ts
Comment (petebacondarwin) [reply packages/wrangler/src/cli.ts:14]: > They'd still be able to see it in the bundle

Doh! Of course. But it would be nice not to have this constant hidden inside the code here, IMO. Perhaps not a build time ENV but at least pulled out into a `const` somewhere obvious.

Comment (petebacondarwin) [reply packages/wrangler/src/cli.ts:14]: Regarding opt-out. The Angular CLI uses a postinstall hook: https://github.com/angular/angular-cli/blob/b3d7080147f755d578d2ee5611437aed9c88f098/packages/angular/cli/bin/postinstall/analytics-prompt.js#L15

Comment (petebacondarwin) [reply packages/wrangler/src/cli.ts:14]: It then has two levels of config: project (local) and global. For the project level, the equivalent would be to store it in the wrangler.toml. For the global then we would could define a file like `~/.wrangler.toml`.

File: packages/wrangler/scripts/bundle.ts
Comment (JacobMGEvans) [reply packages/wrangler/scripts/bundle.ts:25]: I needed to be able to control what the actual URL was that it had in the comment. I will check if it works.

Comment (petebacondarwin) [reply packages/wrangler/scripts/bundle.ts:25]: The `/` should not be needed. SourceMap links via URLs are always relative to the file containing them. So

```
//# sourceMappingURL=cli.js.map
```

"should" be correct.

Can you demo the problem?

Comment (JacobMGEvans) [reply packages/wrangler/scripts/bundle.ts:25]: > The `/` should not be needed. SourceMap links via URLs are always relative to the file containing them. So

It is in this case as far as I have been able to determine.

> ```
> //# sourceMappingURL=cli.js.map
> ```
> 
> "should" be correct.
> 
> Can you demo the problem?

Yes we can discuss in our next meeting.

File: .github/sourcemap-url.js
Comment (petebacondarwin) [inline .github/sourcemap-url.js:6]:
```suggestion
  .readFileSync("./packages/wrangler/wrangler-dist/cli.js", "utf-8");
```

File: .github/sourcemap-url.js
Comment (petebacondarwin) [inline .github/sourcemap-url.js:17]:
```suggestion
else {
```

File: .github/sourcemap-url.js
Comment (petebacondarwin) [inline .github/sourcemap-url.js:5]: Can you pull this string out into a constant, since it is used multiple times in this file.
(It could even be grabbed from the command line args so that it can be configured in the action...?)

File: .github/workflows/release.yml
Comment (petebacondarwin) [inline .github/workflows/release.yml:39]: I think we really need a big comment, either here or in `sourcemap-url.js` explaining why this is necessary.

File: .changeset/sixty-dragons-grab.md
Comment (petebacondarwin) [inline .changeset/sixty-dragons-grab.md:6]:
```suggestion
Sentry will provide insights to errors and debugging potentially allowing us 
```
??

File: .changeset/sixty-dragons-grab.md
Comment (petebacondarwin) [inline .changeset/sixty-dragons-grab.md:7]: There is the slight difference that Wrangler1 sends messages via a Worker proxy, yes?

File: .github/workflows/release.yml
Comment (JacobMGEvans) [inline .github/workflows/release.yml:59]: Tested these scripts locally on a fork where I was running the Actions 
Screen Shot 2022-01-24 at 2 52 34 PM

File: .changeset/sixty-dragons-grab.md
Comment (JacobMGEvans) [reply .changeset/sixty-dragons-grab.md:7]: Correct, should that be mentioned here as an implementation difference?

File: packages/wrangler/src/cli.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/cli.ts:36]: To the credit of the code quality, it is actually quite difficult to create scenarios that are not caught and/or handled gracefully.

File: .github/sourcemap-url.js
Comment (JacobMGEvans) [reply .github/sourcemap-url.js]: > It could even be grabbed from the command line args so that it can be configured in the action...?

I think for this we would need to turn the script into an "Action" and pull in the libraries for that. I could look into it though.

File: packages/wrangler/scripts/bundle.ts
Comment (JacobMGEvans) [reply packages/wrangler/scripts/bundle.ts]: Handling the SourceMap scripting logic in GitHub Actions

File: .changeset/sixty-dragons-grab.md
Comment (petebacondarwin) [reply .changeset/sixty-dragons-grab.md:7]: Perhaps not mention Wrangler 1 at all - since our approach here is not really aiming to replicate what Wrangler 1 does.

Comment (JacobMGEvans) [reply .changeset/sixty-dragons-grab.md]: Improved Changeset information

File: packages/wrangler/src/__tests__/jest.setup.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/jest.setup.ts:6]: By the way, we should probably leave a comment here explaining that in general we don't want to run any sentry stuff in tests.

File: packages/wrangler/package.json
Comment (JacobMGEvans) [inline packages/wrangler/package.json:44]: This is a stopgap. Should reimplement with Ink or find an alternative that uses Ink

File: packages/wrangler/src/__tests__/jest.setup.ts
Comment (JacobMGEvans) [inline packages/wrangler/src/__tests__/jest.setup.ts:16]: Modified this from `.mock()` to `.setMock` which fixed an issue where the fetch was throwing an `isntanceof` Type Error in the logout test using the jest-fetch-mock

Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/jest.setup.ts:16]: Screen Shot 2022-02-07 at 11 38 27 AM

Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/jest.setup.ts:16]: This could need to be changed the mock to handle `undici` and possible replace & remove `jest-fetch-mock` if it cannot handle the `undici` fetch implementation.

File: packages/wrangler/package.json
Comment (petebacondarwin) [reply packages/wrangler/package.json:44]: I think these dependencies can be moved to devDeps since we will be bundling them into the wrangler distributable, right?

File: packages/wrangler/src/__tests__/jest.setup.ts
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/jest.setup.ts:16]: I thought we had already moved over to undici and were no longer using `node-fetch`??

File: packages/wrangler/src/__tests__/sentry.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/sentry.test.ts:24]: NIT: can we have blank lines between these blocks (and the tests).

File: packages/wrangler/src/__tests__/sentry.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/sentry.test.ts:29]: NIT: for consistency all the other tests use `it()` rather than `test()`.

File: packages/wrangler/src/index.tsx
Comment (petebacondarwin) [inline packages/wrangler/src/index.tsx:2103]: NIT:
Being really pedantic and overly cautious here, but how about we do not mention sentry outside of the file where it is implemented? This would make it easier to replace the underlying implementation if we chose to. But also the developer reading the code here doesn't necessarily know what Sentry is, and so this name is not that description. For example, it could be something like `reportError()` or similar.

Taking this to completion, the `sentry.ts` file would be renamed to something more generic, like `reporting.ts`(??); and `Sentry.init()` call would be moved to this `reporting.ts` and exported as something like `initReporting()`.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Some more comments I just thought of. This is looking really good.

File: packages/wrangler/src/__tests__/jest.setup.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/jest.setup.ts:16]: Might have been a rebase that I did and that got lost in the jungle.

File: packages/wrangler/src/reporting.ts
Comment (Electroid) [inline packages/wrangler/src/reporting.ts:73]:
```suggestion
      message: "Would you like to submit a report when an error occurs?",
```

**Review by Electroid — CHANGES_REQUESTED** [review]

File: .changeset/sixty-dragons-grab.md
Comment (petebacondarwin) [inline .changeset/sixty-dragons-grab.md:24]:
```suggestion
  will be added to default.toml with a datetime property for future update checks.
```

File: packages/wrangler/src/reporting.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/reporting.ts]: Done.  👍🏻

File: .github/workflows/release.yml
Comment (petebacondarwin) [inline .github/workflows/release.yml:58]:
```suggestion
      # In the Sentry project there is a concept of a filesystem which if they are all referencing root can be used to determine the correct location of the sourcemap.
```

File: .github/workflows/release.yml
Comment (petebacondarwin) [inline .github/workflows/release.yml:54]: Should we not be getting this version from the local `package.json`?

File: packages/wrangler/src/__tests__/sentry.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/sentry.test.ts:18]: Can you add one more test that checks that Sentry is not called if `isTTY` is false?

**Review by petebacondarwin — APPROVED** [review]
Approving with a couple of suggestions.

File: .github/workflows/release.yml
Comment (JacobMGEvans) [reply .github/workflows/release.yml:54]: This happens after the changeset does a release so it is the latest "released" version to pass to Sentry

Comment (JacobMGEvans) [reply .github/workflows/release.yml:54]: Done.

File: packages/wrangler/src/__tests__/sentry.test.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/sentry.test.ts:18]: Done. 👍🏻 
https://github.com/cloudflare/wrangler2/pull/227/files#diff-530006b680b300438ce683119071a6c83caeff8ddd876b9f886efb7e172e24b7R134-R153

File: .github/sourcemap-url.js
Comment (threepointone) [inline .github/sourcemap-url.js:15]: does this condition ever trigger? we always add a sourcemap.

File: .github/workflows/release.yml
Comment (threepointone) [reply .github/workflows/release.yml:54]: This gets triggered for every commit that lands.

File: package.json
Comment (threepointone) [inline package.json:17]: why did we add this to the root package.json? it's also in packages/wrangler/package.json

File: packages/wrangler/src/reporting.ts
Comment (threepointone) [inline packages/wrangler/src/reporting.ts:70]: why didn't we use ink here btw?

File: packages/wrangler/package.json
Comment (threepointone) [inline packages/wrangler/package.json:39]: we're bundling the packages, so these should probably in devDependencies, yes?

**Comment by petebacondarwin** [comment]
Hmm... this is not the greatest UI: when I run a wrangler command with the wrong options:

```
Not enough non-option arguments: got 0, need at least 1
✔ Would you like to submit a report when an error occurs? › Always
CommandLineArgsError: Not enough non-option arguments: got 0, need at least 1
    at /Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:142668:15
    at Object.fail (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:131057:11)
    at Object.positionalCount (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:131827:14)
    at CommandInstance.populatePositionals (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:130823:21)
    at CommandInstance.applyMiddlewareAndGetResult (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:130780:28)
    at CommandInstance.runCommand (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:130728:212)
    at YargsInstance.[runYargsParserAndExecuteCommands] (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:133422:91)
    at YargsInstance.parse (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:132773:59)
    at main (/Users/pbacondarwin/dev/wrangler2/packages/wrangler/wrangler-dist/cli.js:143872:20)
Waiting for the debugger to disconnect...
```

**Comment by petebacondarwin** [comment]
By the way, we should also (if the user specifies `Always` or `No`) tell the user where the user config file is that we have written to. (E.g. `~/.wrangler/config/default.toml`)

File: package.json
Comment (JacobMGEvans) [reply package.json:17]: Unintentional. I will make a PR to remove

File: packages/wrangler/src/reporting.ts
Comment (threepointone) [reply packages/wrangler/src/reporting.ts:70]: Pete explained  to me that this is temporary till we figure out how to do this well with ink, no problem.

File: .github/workflows/release.yml
Comment (threepointone) [inline .github/workflows/release.yml:62]: ^ Bumping this up, I assume we don't want to run this on every commit?

**Comment by threepointone** [comment]
Filed an issue for the UX https://github.com/cloudflare/wrangler2/issues/444

**Pull request merged**
